### PR TITLE
Support Adding Redirects to Projects and Workspaces

### LINKED ISSUE #1848: Support Redirections
We sometimes have to move workspaces for a variety of reasons.  This can be across an org boundary at times.  This breaks the original URL which feels wrong given our platforms ethos on working in the open.

It's very likely we'll need to move other objects in the future (Projects is the obvious one).  So the following design allows for that.

We should add a table for redirections with an FK to the instance it is a redirection for, a field holding the old URL, and metadata for when&who created/updated/deleted the redirection record, eg

    class Redirect(models.Model):
        workspace = models.ForeignKey("Workspace", …)

        old_url = models.TextField()   # validate this is not empty

        created_at = models.DateTimeField(default=timezone.now)
        created_by = models.ForeignKey("User", …)

        updated_at = models.DateTimeField(auto_add=True)

        deleted_at = models.DateTimeField(null=True)
        deleted_by = models.ForeignKey("User", null=True, …)

While we could use Generic Foreign Keys they've given me so many issues in the past that I think multiple, mutually exclusive/only-1-not-null FKs is fine here.  We can enforce the mutual exclusivity with a `Constraint` in the Meta class.

We should make rows unique on their FK relation.  Since the constraint means we can only ever have one not-null FK we should be able to use `unique_together` or another `Constraint`.

When a relevant View loads an object it can now `select_related()` redirects for it.  If the current URL doesn't match the object's `get_absolute_url()` then it can return a 301 to it.

We likely don't want to keep redirections around forever so the `created_at` field means we can run a cronjob (via Django extensions' jobs framework) to clear out old ones.  Maybe starting with 1 year?

With the table and view additions in place we now need a place to make these moves.  The Staff Area is the obvious place for this, following our standard list and detail views.  A filter for the type of object with redirects would likely be useful.  The home page should include redirects for matching objects of a search result, eg searching `age` might return a Workspace, and should return a redirect linked to that particular Workspace instance too.

### PR DESCRIPTION
**Pull request by ghickman**
This add redirect support, initially just for Projects and Workspaces because we know we need those.  The Redirect model has been created such that further objects could be support with what I hope is relative ease (the expected pain point will be the constraints).

This supports  N redirects for a given target object since the destination of a redirect is based on the targets `.get_absolute_url()`, so will always point to the current location.

Redirects are created when editing either the parent object of a Project (org) or a Workspace (project), and will show up in the detail pages for either of those.

The middleware catches possible redirects so views don't have to care about them.  Both Projects and Workspaces have child objects which are affected by a redirect (eg the Released outputs page of a Workspace is affected by the Org of a Project changing), so the middleware also searches for prefix matches to handle those too.

Fixes #1848

### CODE REVIEW

**Review by bloodearnest — APPROVED** [review]
Left a performance query that could be addressed in a follow on PR if needed.

How can we ensure a base set of project/workspace redirects is set up? Do we want a db fixture?

File: redirects/middleware.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    def __call__(self, request):
+        exemptions = (
+            "/api",
+            "/staff",
+        )
+        if request.path.startswith(exemptions):
+            return self.get_response(request)
+
+        # TODO: could we use the db to look for prefixes of the current URL??
+        redirects = Redirect.objects.all()
```
[the remark this answers is not in the document]
Comment (ghickman) [reply redirects/middleware.py:29]: TIL you can do `value LIKE field`!

**Comment by bloodearnest** [comment]
> > How can we ensure a base set of project/workspace redirects is set up?
> 
> What do we want this for?
> 
> @bloodearnest – updated the query, that's much nicer +1

So we make sure that local dev has the same redirects as prod

**Comment by ghickman** [comment]
@bloodearnest – we have regular db dumps available to get redirects if we need them.  They're currently generated from user actions so it's not a pre-configured list.

**Pull request merged**